### PR TITLE
use mutex for channels instead of thread data

### PR DESCRIPTION
**Pull request by Benolavi**

### CODE REVIEW

File: test/pubsub.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+require_relative 'config'
```
Comment (quantumlicht) [inline test/pubsub.rb]: Can you also add unit tests? These kind of tests won't be that helpful when we integration continuous integration. Also they don't state what is expected and have no doc, so they are unlikely to be used by anyone.

File: test/pubsub.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
```
Comment (ralfkrauklis) [inline test/pubsub.rb]: Can you implement some rspec / minitests that cover the code to ensure that nothing is broken?

File: lib/videri/message_bus.rb [hunk 3]
```diff
@@ -70,30 +72,36 @@ def self.subscribe(topic, key, queue_opts = {})
...
   end
 
   private
+  def self.mutex
+    @@mutex ||= Mutex.new
+  end
+
   def self.conn
-    return @@conn unless @@conn.nil?
+    @@conn unless @@conn.nil?
```
Comment (Benolavi) [inline lib/videri/message_bus.rb]: ok, good catch

File: lib/videri/message_bus.rb [hunk 4]
```diff
@@ -102,25 +110,26 @@ def self.conn
   def self.close
     @@conn.close unless @@conn.nil?
```
Comment (Benolavi) [inline lib/videri/message_bus.rb]: ok wrapped with empty rescue

File: lib/videri/message_bus.rb [hunk 5]
```diff
@@ -102,25 +110,26 @@ def self.conn
...
     @@conn = nil
+    release_channel
```
Comment (Benolavi) [inline lib/videri/message_bus.rb]: A close connection will clean up all related. As for instantiated objects, the GC will take care of that

File: test/pubsub.rb (hunk 2 above)
Comment (Benolavi) [inline test/pubsub.rb]: yes in the VR-6689

File: lib/videri/message_bus.rb [hunk 6]
```diff
@@ -70,30 +72,36 @@ def self.subscribe(topic, key, queue_opts = {})
...
-      logger.error e.inspect
+    mutex.synchronize do
+      begin
+        payload = params[:payload].merge(:timestamp => DateTime.now, :key => key, :host => host)
+        exchange(topic).publish payload.to_json, :routing_key => key
+
+      rescue => e
+        logger.warn e.inspect
+        release_channel
+        sleep 1
```
Comment (Benolavi) [inline lib/videri/message_bus.rb]: Nope, just an attempt to recover nicely.

Comment (Benolavi) [inline lib/videri/message_bus.rb]: Should we just fail on the first rescue ?

File: lib/videri/message_bus.rb [hunk 7]
```diff
@@ -70,30 +72,36 @@ def self.subscribe(topic, key, queue_opts = {})
...
-
-    rescue => e
-      logger.warn e.inspect
-      cleanup topic
-      sleep 1
-      retry if (retries += 1) < 4
-      logger.error e.inspect
+    mutex.synchronize do
+      begin
+        payload = params[:payload].merge(:timestamp => DateTime.now, :key => key, :host => host)
```
Comment (Benolavi) [inline lib/videri/message_bus.rb]: ok, agreed

File: test/pubsub.rb (hunk 1 above)
Comment (Benolavi) [inline test/pubsub.rb]: Forthcoming

File: lib/videri/message_bus.rb [hunk 8]
```diff
@@ -102,25 +110,26 @@ def self.conn
   def self.close
```
Comment (Benolavi) [inline lib/videri/message_bus.rb]: close is reserved for future use. Right now it only affects publishers.

File: test/Gemfile [hunk 9]
```diff
@@ -1,7 +1,9 @@
 #!/usr/bin/env bash
 source 'http://rubygems.org'
-source 'http://software:Videri123@gems.videri.com/'
 
 ruby '2.1.5'
 
-gem 'videri-messagebus'
```
Comment (Benolavi) [inline test/Gemfile]: Testing gem should always has to be done with local files, not gem from repos,

File: lib/videri/message_bus.rb [hunk 10]
```diff
@@ -1,23 +1,26 @@
+require 'json'
```
Comment (Benolavi) [inline lib/videri/message_bus.rb]: Was missing this require. The gem must be self contained and not assume users to require additional depends.

File: README.rdoc [hunk 11]
```diff
@@ -21,6 +21,6 @@
     rake (Rakefile) 
     
 * How to run tests
-   ruby ingest_subscriber [dev|qa] should receive messages from the system.
-   
+   rake test (all test run on dev RabbitMQ)
+
```
Comment (Benolavi) [inline README.rdoc]: Using minitest. Only MessageBus pubsub is part of the tests for now.

File: Rakefile [hunk 12]
```diff
@@ -1,7 +1,13 @@
...
 task :build_gem do
   system %{gem build videri-messagebus.gemspec}
-end
\ No newline at end of file
+end
+
+Rake::TestTask.new do |t|
+  t.test_files = FileList['test/*_test.rb']
+  t.warning = false
+end
```
Comment (Benolavi) [inline Rakefile]: Add only files to be part o minit test. (rake test)

**Pull request merged**